### PR TITLE
[WIP] Fix TIE phase retrieval

### PR DESCRIPTION
**Pull request by tfarago**
This is the start of fixing the phase retrieval which tries to make the results quantitatively correct.

### CODE REVIEW

File: tofu/preprocess.py [hunk 1]
```diff
@@ -157,8 +158,25 @@ def create_phase_retrieval_pipeline(args, graph, processing_node=None):
...
+        if args.fix_nan_and_inf:
+            expression += '(isinf (v) || isnan (v) || (v <= 0)) ? 0.0f :'
+        expression += '- log (v)'
+        if args.delta is not None:
+            import numpy as np
+            lam = 6.62606896e-34 * 299792458 / (args.energy * 1.60217733e-16)
+            # Compute mju from the fact that beta = 10^-regularization_rate * delta
+            # and mju = 4 * Pi * beta / lambda
+            mju = 4 * np.pi * 10 ** -args.regularization_rate * args.delta / lam
+            expression = '-log ({} * v) * {}'.format(2 / 10 ** args.regularization_rate, 1 / mju)
```
Comment (tfarago) [inline tofu/preprocess.py]: The first parameter of `format` should rather be fixed in `ufo-filters`. I will wait what tomorrow's discussion brings and might shift it there.

**Comment by tfarago** [comment]
This is probably going to be interesting for @milkdrop and @lucito

File: tofu/preprocess.py (hunk 1 above)
Comment (tfarago) [reply tofu/preprocess.py]: The code as it is now gives the projected thickness which can be directly plugged into e.g. tomographic reconstruction.

Comment (tfarago) [reply tofu/preprocess.py]: @matze don't mind the `expression +=` vs `expression = ` inconsistency, the code is anyway going to change tomorrow. This is just to show the procedure which needs to be carried out to get the projected thickness.

**Comment by milkdrop** [comment]
Why the following is conditional on fix_nan_an_inf parameter? 
if args.fix_nan_and_inf:
        expression += '(isinf (v) || isnan (v) || (v <= 0)) ? 0.0f :'
expression += '- log (v)'
Logically, assume that one did not specify (or has forgotten to specify) fix-nan-and-inf. Then, the result of TIE will very likely have plenty of bad pixels, while output of flatcorrect might have a couple of bad pixels. Now I can imagine that in a particular case someone would like to get flatcorrected images "as is" without any post-filtering. Then he can just skip the fix-nan-and-inf parameter. But I fail to imagine a situation when someone would like to risk applying a "log" to an image that might potentially contain infs, nans, or negative numbers. If you remove "if" there, then the output of TIE will still have only well-defined numbers then even if fix-nan-and-inf is not defined.

**Comment by tfarago** [comment]
Someone applying it to e.g. simulated data when one is certain everything is OK. Also, it is consistent with the usage of the flag for the flat correction. What does @matze think?

**Comment by matze** [comment]
> Someone applying it to e.g. simulated data when one is certain everything is OK.

Maybe, but now you are branching a lot :smile:

**Comment by tfarago** [comment]
OK, I guess I will get rid of that when I am on it. Still waiting for one guy, I will get back to you on this later.

**Comment by tfarago** [comment]
This PR was merged with 42a432ca70ec0515b1849e59a1701febe74db2aa, and #84 supersedes it, so I will close.

**Pull request closed without merging**